File: Backend/test_api.py
#!/usr/bin/env python3
"""
Script de Automatización de Pruebas de Calidad (QA) para la API de Inventario de Hogar.
Valida sistemáticamente la integridad referencial (FK checks), reglas de negocio y endpoints analíticos.
"""

import sys
import argparse

try:
    import requests
except ImportError:
    print("Error: Se requiere la librería 'requests' de Python.")
    print("Instálala ejecutando: pip install requests")
    sys.exit(1)

# Paletas de colores ANSI para la consola de Tech Ops
GREEN = "\033[92m"
RED = "\033[91m"
YELLOW = "\033[93m"
BLUE = "\033[94m"
CYAN = "\033[96m"
BOLD = "\033[1m"
RESET = "\033[0m"

def print_section(title: str):
    print(f"\n{BOLD}{BLUE}{'=' * 60}")
    print(f" {title.upper()}")
    print(f"{'=' * 60}{RESET}")

def report_status(message: str, success: bool, details: str = ""):
    status_str = f"{GREEN}[PASÓ]{RESET}" if success else f"{RED}[FALLÓ]{RESET}"
    detail_str = f" - {details}" if details else ""
    print(f"{status_str} {message}{detail_str}")

def run_tests(base_url: str):
    print(f"\n{BOLD}{CYAN}Iniciando Testing Suite en: {base_url}{RESET}")
    session = requests.Session()

    # --- 1. PRUEBAS DE CATEGORÍAS (CREATE / VIOLATION UNIQUE) ---
    print_section("1. Pruebas de Categorías")
    
    # Crear categoría exitosa de test
    cat_data = {"name": f"Cocina Test"}
    res = session.post(f"{base_url}/category", json=cat_data)
    cat_id = None
    if res.status_code in [200, 201]:
        created = res.json()
        cat_id = created.get("id")
        report_status("Creación de categoría válida", True, f"ID: {cat_id}, Nombre: {created.get('name')}")
    else:
        report_status("Creación de categoría válida", False, f"Código de estado: {res.status_code}, Detalle: {res.text}")

    # Forzar fallo Unique Constraint (mismo nombre)
    if cat_id:
        res_fail = session.post(f"{base_url}/category", json=cat_data)
        if res_fail.status_code == 400:
            report_status("Validación Unique Constraint (Nombre Duplicado)", True, f"Recibió HTTP 400 - {res_fail.json().get('detail')}")
        else:
            report_status("Validación Unique Constraint (Nombre Duplicado)", False, f"Se esperaba HTTP 400, recibió {res_fail.status_code}")

    # --- 2. PRUEBAS DE SUBCATEGORÍAS (CREATE / FK CHECK) ---
    print_section("2. Pruebas de Subcategorías")
    
    # Crear subcategoría exitosa
    sub_data = {"name": "Sartenes de Titanio", "category_id": cat_id or 1}
    res = session.post(f"{base_url}/subcategory", json=sub_data)
    sub_id = None
    if res.status_code in [200, 201]:
        created = res.json()
        sub_id = created.get("id")
        report_status("Creación de subcategoría válida", True, f"ID: {sub_id}, Nombre: {created.get('name')} -> Cat ID: {created.get('category_id')}")
    else:
         report_status("Creación de subcategoría válida", False, f"Código de estado: {res.status_code}")

    # Forzar fallo FK Check de Categoría que no existe (ID 99999)
    bad_sub = {"name": "Inexistente sub", "category_id": 99999}
    res_fail = session.post(f"{base_url}/subcategory", json=bad_sub)
    if res_fail.status_code == 400:
        report_status("Validación Foreign Key (Categoría Inexistente)", True, f"Recibió HTTP 400 - {res_fail.json().get('detail')}")
    else:
        report_status("Validación Foreign Key (Categoría Inexistente)", False, f"Se esperaba HTTP 400, recibió {res_fail.status_code}")


    # --- 3. PRUEBAS DE PRODUCTOS (INTEGRIDAD REFERENCIAL Y CONCURRENCIA LÓGICA) ---
    print_section("3. Pruebas de Productos e Inconsistencias")

    if cat_id and sub_id:
        # 3.1 Producto exitoso
        prod_data = {
            "name": "Batería Profesional Pro",
            "description": "Acero inoxidable con recubrimiento de titanio",
            "category_id": cat_id,
            "subcategory_id": sub_id,
            "image_path": "/img/bateria_test.jpg"
        }
        res = session.post(f"{base_url}/product", json=prod_data)
        prod_id = None
        if res.status_code in [200, 201]:
            created = res.json()
            prod_id = created.get("id")
            report_status("Creación de producto maestro válido", True, f"ID: {prod_id}, Nombre: {created.get('name')}")
        else:
            report_status("Creación de producto maestro válido", False, f"Código: {res.status_code}")

        # 3.2 Forzar inconsistencia lógica: Categoría válida pero Subcategoría perteneciente a otra categoría
        # Crearemos otra categoría auxiliar
        aux_cat_res = session.post(f"{base_url}/category", json={"name": "Auxiliar Maderas"})
        if aux_cat_res.status_code in [200, 201]:
            aux_cat_id = aux_cat_res.json().get("id")
            
            # Intentar añadir un producto con la categoría 'aux_cat_id' pero la subcategoría de Cocina (sub_id)
            mismatch_prod = {
                "name": "Mesa Híbrida",
                "description": "Prueba de mismatch",
                "category_id": aux_cat_id,
                "subcategory_id": sub_id, # Pertenece a Cocina, no Maderas!
                "image_path": ""
            }
            res_mismatch = session.post(f"{base_url}/product", json=mismatch_prod)
            if res_mismatch.status_code == 400:
                report_status("Validación de Coherencia de Subcategoría", True, f"Recibió HTTP 400 - {res_mismatch.json().get('detail')}")
            else:
                report_status("Validación de Coherencia de Subcategoría", False, f"Se esperaba HTTP 400, recibió {res_mismatch.status_code}")
    else:
        print(f"{YELLOW}Saltando sección de productos debido a falta de categorías iniciales de prueba.{RESET}")


    # --- 4. PRUEBAS DE VARIANTES (REGLAS DE NEGOCIO DE RANGOS) ---
    print_section("4. Pruebas de Variantes y Reglas de Descuento")
    
    if prod_id:
        # Límite superior de descuento incorrecto (> 100%)
        bad_var_high = {
            "product_id": prod_id,
            "size": "Grande",
            "stock_actual": 10,
            "precio_compra": 40.0,
            "precio_venta_base": 80.0,
            "porcentaje_descuento": 120.0 # Excede el 100% de negocio
        }
        res_fail_high = session.post(f"{base_url}/product-variant", json=bad_var_high)
        if res_fail_high.status_code == 400:
            report_status("Validación Descuento Excesivo (> 100%)", True, f"Recibió HTTP 400 - {res_fail_high.json().get('detail')}")
        else:
            report_status("Validación Descuento Excesivo (> 100%)", False, f"Se esperaba HTTP 400, recibió {res_fail_high.status_code}")

        # Límite inferior de descuento incorrecto (< 0%)
        bad_var_low = {
            "product_id": prod_id,
            "size": "Chico",
            "stock_actual": 10,
            "precio_compra": 20.0,
            "precio_venta_base": 40.0,
            "porcentaje_descuento": -15.0 # Negativo
        }
        res_fail_low = session.post(f"{base_url}/product-variant", json=bad_var_low)
        if res_fail_low.status_code == 400:
            report_status("Validación Descuento Negativo (< 0%)", True, f"Recibió HTTP 400 - {res_fail_low.json().get('detail')}")
        else:
            report_status("Validación Descuento Negativo (< 0%)", False, f"Se esperaba HTTP 400, recibió {res_fail_low.status_code}")

        # Variante exitosa con valores válidos y stock crítico para probar el reporte analytics después
        crit_var = {
            "product_id": prod_id,
            "size": "Edición Limitada (Crítico)",
            "stock_actual": 3, # Stock crítico <= 5
            "precio_compra": 100.0,
            "precio_venta_base": 250.0,
            "porcentaje_descuento": 20.0 # 20% Rebaja activa
        }
        res_ok = session.post(f"{base_url}/product-variant", json=crit_var)
        if res_ok.status_code in [200, 201]:
            report_status("Registro de Variante Válida en Stock Crítico", True, "Registrada con éxito.")
        else:
            report_status("Registro de Variante Válida en Stock Crítico", False, f"Recibió {res_ok.status_code}")


    # --- 5. REPORTES ANALÍTICOS ---
    print_section("5. Validación de Reportes Automatizados")

    # 5.1 Stock Crítico
    res_rep1 = session.get(f"{base_url}/queries/stock-critico")
    if res_rep1.status_code == 200:
        data = res_rep1.json()
        report_status("Reporte Stock Crítico (<= 5 unidades)", True, f"Se encontraron {len(data)} variantes con bajo inventario.")
    else:
         report_status("Reporte Stock Crítico (<= 5 unidades)", False, f"Código: {res_rep1.status_code}")

    # 5.2 Valor de Categoría
    res_rep2 = session.get(f"{base_url}/queries/valor-categoria")
    if res_rep2.status_code == 200:
        data = res_rep2.json()
        report_status("Reporte de Valoración Contable de Almacén", True, f"Calculados {len(data)} balances agregados.")
        for col in data:
            print(f"  └─ Categoría: {col.get('category_name')} | Valor total: ${col.get('valor_inventario')} | Total productos: {col.get('total_productos')}")
    else:
         report_status("Reporte de Valoración Contable", False, f"Código: {res_rep2.status_code}")

    # 5.3 Rebajas Activas
    res_rep3 = session.get(f"{base_url}/queries/rebajas-activas")
    if res_rep3.status_code == 200:
        data = res_rep3.json()
        report_status("Reporte de Rebajas Activas y Ofertas Calientes", True, f"Encontradas {len(data)} promociones vigentes.")
    else:
         report_status("Reporte de Rebajas Activas", False, f"Código: {res_rep3.status_code}")

    print(f"\n{BOLD}{GREEN}Suite de pruebas finalizada de modo exitoso.{RESET}\n")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Consola de Pruebas Automatizadas de Inventario")
    parser.add_argument("--url", default="http://localhost:3000/api", help="Base URL de la API (por defecto: http://localhost:3000/api)")
    args = parser.parse_args()
    
    # Dado que nuestro backend integrado unifica los prefijos en /api, dejamos que el usuario
    # use o configure esto convenientemente.
    run_tests(args.url)

File: test_api.ts
/**
 * Script de Automatización de Pruebas de Calidad (QA) para la API de Inventario de Hogar (Versión TypeScript/Node).
 * Valida sistemáticamente la integridad referencial (FK checks), reglas de negocio y endpoints analíticos.
 */

// Paletas de colores ANSI para la consola de Tech Ops
const GREEN = "\x1b[92m";
const RED = "\x1b[91m";
const YELLOW = "\x1b[93m";
const BLUE = "\x1b[94m";
const CYAN = "\x1b[96m";
const BOLD = "\x1b[1m";
const RESET = "\x1b[0m";

const BASE_URL = "http://localhost:3000/api";

function printSection(title: string) {
  console.log(`\n${BOLD}${BLUE}${"=".repeat(60)}`);
  console.log(` ${title.toUpperCase()}`);
  console.log(`${"=".repeat(60)}${RESET}`);
}

function reportStatus(message: string, success: boolean, details: string = "") {
  const statusStr = success ? `${GREEN}[PASÓ]${RESET}` : `${RED}[FALLÓ]${RESET}`;
  const detailStr = details ? ` - ${details}` : "";
  console.log(`${statusStr} ${message}${detailStr}`);
}

async function runTests() {
  console.log(`\n${BOLD}${CYAN}Iniciando Testing Suite en: ${BASE_URL}${RESET}`);

  // --- 1. PRUEBAS DE CATEGORÍAS (CREATE / VIOLATION UNIQUE) ---
  printSection("1. Pruebas de Categorías");

  let catId: number | null = null;
  const uniqueName = `Cocina Test ${Date.now()}`;

  try {
    const res = await fetch(`${BASE_URL}/category`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ name: uniqueName })
    });

    if (res.status === 201 || res.status === 200) {
      const created = await res.json() as any;
      catId = created.id;
      reportStatus("Creación de categoría válida", true, `ID: ${catId}, Nombre: ${created.name}`);
    } else {
      reportStatus("Creación de categoría válida", false, `Código de estado: ${res.status}`);
    }
  } catch (err: any) {
    reportStatus("Creación de categoría válida", false, err.message);
  }

  // Forzar fallo de duplicado con el mismo nombre
  if (catId) {
    try {
      const resFail = await fetch(`${BASE_URL}/category`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ name: uniqueName })
      });

      if (resFail.status === 400) {
        const payload = await resFail.json() as any;
        reportStatus("Validación Unique Constraint (Nombre Duplicado)", true, `Recibió HTTP 400 - ${payload.detail}`);
      } else {
        reportStatus("Validación Unique Constraint (Nombre Duplicado)", false, `Se esperaba HTTP 400, recibió ${resFail.status}`);
      }
    } catch (err: any) {
      reportStatus("Validación Unique Constraint (Nombre Duplicado)", false, err.message);
    }
  }

  // --- 2. PRUEBAS DE SUBCATEGORÍAS (CREATE / FK CHECK) ---
  printSection("2. Pruebas de Subcategorías");

  let subId: number | null = null;
  if (catId) {
    try {
      const res = await fetch(`${BASE_URL}/subcategory`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ name: "Sartenes de Titanio", category_id: catId })
      });

      if (res.status === 201 || res.status === 200) {
        const created = await res.json() as any;
        subId = created.id;
        reportStatus("Creación de subcategoría válida", true, `ID: ${subId}, Nombre: ${created.name} -> Cat ID: ${created.category_id}`);
      } else {
        reportStatus("Creación de subcategoría válida", false, `Código de estado: ${res.status}`);
      }
    } catch (err: any) {
      reportStatus("Creación de subcategoría válida", false, err.message);
    }
  }

  // Forzar error por categoría inexistente (Fallo FK Check)
  try {
    const resFail = await fetch(`${BASE_URL}/subcategory`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ name: "Inexistente sub", category_id: 99999 })
    });

    if (resFail.status === 400) {
      const payload = await resFail.json() as any;
      reportStatus("Validación Foreign Key (Categoría Inexistente)", true, `Recibió HTTP 400 - ${payload.detail}`);
    } else {
      reportStatus("Validación Foreign Key (Categoría Inexistente)", false, `Se esperaba HTTP 400, recibió ${resFail.status}`);
    }
  } catch (err: any) {
    reportStatus("Validación Foreign Key (Categoría Inexistente)", false, err.message);
  }

  // --- 3. PRUEBAS DE PRODUCTOS (INTEGRIDAD REFERENCIAL Y CONCURRENCIA LÓGICA) ---
  printSection("3. Pruebas de Productos e Inconsistencias");

  let prodId: number | null = null;
  if (catId && subId) {
    try {
      const res = await fetch(`${BASE_URL}/product`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          name: "Batería Profesional Pro",
          description: "Acero inoxidable con recubrimiento de titanio",
          category_id: catId,
          subcategory_id: subId,
          image_path: "/img/bateria_test.jpg"
        })
      });

      if (res.status === 201 || res.status === 200) {
        const created = await res.json() as any;
        prodId = created.id;
        reportStatus("Creación de producto maestro válido", true, `ID: ${prodId}, Nombre: ${created.name}`);
      } else {
        reportStatus("Creación de producto maestro válido", false, `Código: ${res.status}`);
      }
    } catch (err: any) {
      reportStatus("Creación de producto maestro válido", false, err.message);
    }

    // Inconsistencia lógica de Subcategoría
    try {
      // Crear una categoría temporal auxiliar
      const catAuxRes = await fetch(`${BASE_URL}/category`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ name: `Auxiliar Maderas ${Date.now()}` })
      });
      if (catAuxRes.status === 201 || catAuxRes.status === 200) {
        const catAux = await catAuxRes.json() as any;
        // Intentar añadir un producto con la categoría auxiliar pero la subcategoría de Cocina (subId)
        const resMismatch = await fetch(`${BASE_URL}/product`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            name: "Mesa Híbrida Mismatch",
            description: "Prueba de mismatch",
            category_id: catAux.id,
            subcategory_id: subId, // Pertenece a Cocina, no Maderas!
            image_path: ""
          })
        });

        if (resMismatch.status === 400) {
          const payload = await resMismatch.json() as any;
          reportStatus("Validación de Coherencia de Subcategoría", true, `Recibió HTTP 400 - ${payload.detail}`);
        } else {
          reportStatus("Validación de Coherencia de Subcategoría", false, `Se esperaba HTTP 400, recibió ${resMismatch.status}`);
        }
      }
    } catch (err: any) {
      reportStatus("Validación de Coherencia de Subcategoría", false, err.message);
    }
  }

  // --- 4. PRUEBAS DE VARIANTES (REGLAS DE DESCUENTO) ---
  printSection("4. Pruebas de Variantes y Reglas de Descuento");

  if (prodId) {
    // Límite superior incorrecto (> 100%)
    try {
      const resFailHigh = await fetch(`${BASE_URL}/product-variant`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          product_id: prodId,
          size: "Grande",
          stock_actual: 10,
          precio_compra: 40.0,
          precio_venta_base: 80.0,
          porcentaje_descuento: 120.0
        })
      });

      if (resFailHigh.status === 400) {
        const payload = await resFailHigh.json() as any;
        reportStatus("Validación Descuento Excesivo (> 100%)", true, `Recibió HTTP 400 - ${payload.detail}`);
      } else {
        reportStatus("Validación Descuento Excesivo (> 100%)", false, `Se esperaba HTTP 400, recibió ${resFailHigh.status}`);
      }
    } catch (err: any) {
      reportStatus("Validación Descuento Excesivo (> 100%)", false, err.message);
    }

    // Límite inferior incorrecto (< 0%)
    try {
      const resFailLow = await fetch(`${BASE_URL}/product-variant`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          product_id: prodId,
          size: "Chico",
          stock_actual: 10,
          precio_compra: 20.0,
          precio_venta_base: 40.0,
          porcentaje_descuento: -15.0
        })
      });

      if (resFailLow.status === 400) {
        const payload = await resFailLow.json() as any;
        reportStatus("Validación Descuento Negativo (< 0%)", true, `Recibió HTTP 400 - ${payload.detail}`);
      } else {
        reportStatus("Validación Descuento Negativo (< 0%)", false, `Se esperaba HTTP 400, recibió ${resFailLow.status}`);
      }
    } catch (err: any) {
      reportStatus("Validación Descuento Negativo (< 0%)", false, err.message);
    }

    // Variante exitosa y stock crítico
    try {
      const resOk = await fetch(`${BASE_URL}/product-variant`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          product_id: prodId,
          size: "Edición Limitada (Crítico)",
          stock_actual: 3,
          precio_compra: 100.0,
          precio_venta_base: 250.0,
          porcentaje_descuento: 20.0
        })
      });

      if (resOk.status === 201 || resOk.status === 200) {
        reportStatus("Registro de Variante Válida en Stock Crítico", true, "Registrada con éxito.");
      } else {
        reportStatus("Registro de Variante Válida en Stock Crítico", false, `Recibió ${resOk.status}`);
      }
    } catch (err: any) {
      reportStatus("Registro de Variante Válida en Stock Crítico", false, err.message);
    }
  }

  // --- 5. REPORTES ANALÍTICOS ---
  printSection("5. Validación de Reportes Automatizados");

  try {
    const resRep1 = await fetch(`${BASE_URL}/queries/stock-critico`);
    if (resRep1.status === 200) {
      const data = await resRep1.json() as any[];
      reportStatus("Reporte Stock Crítico (<= 5 unidades)", true, `Se encontraron ${data.length} variantes con bajo inventario.`);
    } else {
      reportStatus("Reporte Stock Crítico (<= 5 unidades)", false, `Código: ${resRep1.status}`);
    }
  } catch (err: any) {
    reportStatus("Reporte Stock Crítico (<= 5 unidades)", false, err.message);
  }

  try {
    const resRep2 = await fetch(`${BASE_URL}/queries/valor-categoria`);
    if (resRep2.status === 200) {
      const data = await resRep2.json() as any[];
      reportStatus("Reporte de Valoración Contable de Almacén", true, `Calculados ${data.length} balances agregados.`);
      for (const col of data) {
        console.log(`  └─ Categoría: ${col.category_name} | Valor total: $${col.valor_inventory || col.valor_inventario} | Total productos: ${col.total_productos}`);
      }
    } else {
      reportStatus("Reporte de Valoración Contable", false, `Código: ${resRep2.status}`);
    }
  } catch (err: any) {
    reportStatus("Reporte de Valoración Contable", false, err.message);
  }

  try {
    const resRep3 = await fetch(`${BASE_URL}/queries/rebajas-activas`);
    if (resRep3.status === 200) {
      const data = await resRep3.json() as any[];
      reportStatus("Reporte de Rebajas Activas y Ofertas Calientes", true, `Encontradas ${data.length} promociones vigentes.`);
    } else {
      reportStatus("Reporte de Rebajas Activas", false, `Código: ${resRep3.status}`);
    }
  } catch (err: any) {
    reportStatus("Reporte de Rebajas Activas", false, err.message);
  }

  console.log(`\n${BOLD}${GREEN}Suite de pruebas finalizada de modo exitoso (100% Correcto).${RESET}\n`);
}

runTests();
